### PR TITLE
oauth auth server settings resource

### PR DESCRIPTION
**Pull request by wesleymccollam**
- oauth auth server settings resource
- internal methods
- descriptions added to administrative accounts
- Makefile acceptance test parallelism set from 4 to 1

### CODE REVIEW

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 1]
```diff
@@ -0,0 +1,1167 @@
...
+			"client_secret_retention_period": schema.Int64Attribute{
+				Description: "The length of time in minutes that client secrets will be retained as secondary secrets after secret change. The default value is 0, which will disable secondary client secret retention.",
+				Computed:    true,
+				Optional:    true,
+				PlanModifiers: []planmodifier.Int64{
+					int64planmodifier.UseStateForUnknown(),
+				},
+			},
+			"jwt_secured_authorization_response_mode_lifetime": schema.Int64Attribute{
+				Description: "The lifetime, in seconds, of the JWT Secured authorization response. The default value is 600.",
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:572]: It looks like we know the defaults for a lot of these attributes, so it would be nice to set them in the schema here to have more consistent provider runs across versions.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 2]
```diff
@@ -0,0 +1,1167 @@
...
+	ApprovedAuthorizationDetailAttribute        types.String `tfsdk:"approved_authorization_detail_attribute"`
+	ParReferenceTimeout                         types.Int64  `tfsdk:"par_reference_timeout"`
+	ParReferenceLength                          types.Int64  `tfsdk:"par_reference_length"`
+	ParStatus                                   types.String `tfsdk:"par_status"`
+	ClientSecretRetentionPeriod                 types.Int64  `tfsdk:"client_secret_retention_period"`
+	JwtSecuredAuthorizationResponseModeLifetime types.Int64  `tfsdk:"jwt_secured_authorization_response_mode_lifetime"`
+}
+
+// GetSchema defines the schema for the resource.
+func (r *oauthAuthServerSettingsResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:98]: We don't need to split this into separate methods like the PD provider does if we don't have default_ versions of resources. The same applies to the other methods (read, update, etc.)

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 3]
```diff
@@ -0,0 +1,1167 @@
...
+	if internaltypes.IsDefined(model.Scopes) {
+		scopeElems := model.Scopes.Elements()
+		for _, scopeElem := range scopeElems {
+			scopeElemObjectAttrs := scopeElem.(types.Object)
+			scopeEntryName := scopeElemObjectAttrs.Attributes()["name"].(basetypes.StringValue).ValueString()
+			scopeNames = append(scopeNames, scopeEntryName)
+			scopeEntryIsDynamic := scopeElemObjectAttrs.Attributes()["dynamic"].(basetypes.BoolValue).ValueBool()
+			if scopeEntryIsDynamic {
+				if string(scopeEntryName[0]) != "*" {
+					resp.Diagnostics.AddError("Scope name conflict!", fmt.Sprintf("Scope name \"%s\" must be prefixed, suffixed or both by static part(s) with one and only one dynamic part represented by \"*\"", scopeEntryName))
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:620]: This error message is confusing to me compared to what the code is checking, which is just that the first character is an asterisk.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 4]
```diff
@@ -0,0 +1,1167 @@
...
+	// Test scope names for dynamic true, string must be prepended with *
+	if internaltypes.IsDefined(model.Scopes) {
+		scopeElems := model.Scopes.Elements()
+		for _, scopeElem := range scopeElems {
+			scopeElemObjectAttrs := scopeElem.(types.Object)
+			scopeEntryName := scopeElemObjectAttrs.Attributes()["name"].(basetypes.StringValue).ValueString()
+			scopeNames = append(scopeNames, scopeEntryName)
+			scopeEntryIsDynamic := scopeElemObjectAttrs.Attributes()["dynamic"].(basetypes.BoolValue).ValueBool()
+			if scopeEntryIsDynamic {
+				if string(scopeEntryName[0]) != "*" {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:619]: Add a check for string length >= 1 here to avoid a panic.

Also I'm not sure if this is a safe way to get the first character (rune) of a string in go - it might not handle non-ASCII characters well. We could write some utility function that converts a string to a slice of runes potentially.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 5]
```diff
@@ -0,0 +1,1167 @@
...
+	if internaltypes.IsDefined(model.ExclusiveScopes) {
+		exclusiveScopeElems := model.ExclusiveScopes.Elements()
+		for _, esElem := range exclusiveScopeElems {
+			esElemObjectAttrs := esElem.(types.Object)
+			eScopeEntryName := strings.ToLower(esElemObjectAttrs.Attributes()["name"].(basetypes.StringValue).ValueString())
+			eScopeNames = append(eScopeNames, eScopeEntryName)
+			eScopeEntryIsDynamic := esElemObjectAttrs.Attributes()["dynamic"].(basetypes.BoolValue).ValueBool()
+			if eScopeEntryIsDynamic {
+				if string(eScopeEntryName[0]) != "*" {
+					resp.Diagnostics.AddError("Dynamic scope name conflict!", fmt.Sprintf("Dynamic scope name \"%s\" must be prefixed, suffixed or both by static part(s) with one and only one dynamic part represented by \"*\"", eScopeEntryName))
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:637]: Same comments as the above

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 6]
```diff
@@ -0,0 +1,1167 @@
...
+
+	if internaltypes.IsDefined(plan.Scopes) {
+		err := json.Unmarshal([]byte(internaljson.FromValue(plan.Scopes)), &addRequest.Scopes)
+		if err != nil {
+			return err
+		}
+	}
+
+	if internaltypes.IsDefined(plan.ScopeGroups) {
+		addRequest.ScopeGroups = client.NewAuthorizationServerSettingsWithDefaults().ScopeGroups
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:664]: Why is this needed?

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 7]
```diff
@@ -0,0 +1,1167 @@
...
+	}
+
+	providerCfg := req.ProviderData.(internaltypes.ResourceConfiguration)
+	r.providerConfig = providerCfg.ProviderConfig
+	r.apiClient = providerCfg.ApiClient
+
+}
+
+func readOauthAuthServerSettingsResponse(ctx context.Context, r *client.AuthorizationServerSettings, state *oauthAuthServerSettingsResourceModel) {
+	state.Id = types.StringValue("id")
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:870]: Is there a unique value in the response that would be usable here? Ideally we can have different ids for different resources of the same type.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 8]
```diff
@@ -0,0 +1,1167 @@
...
+	}
+	state.ExclusiveScopeGroups, _ = types.SetValueFrom(ctx, types.ObjectType{AttrTypes: exclusiveScopeGroupAttrTypes}, toStateExclusiveScopeGroups)
+
+	// state.PersistentGrantContract
+	getPersistentGrantContract := r.GetPersistentGrantContract()
+	// Build core attributes
+	nameAttributeType := map[string]attr.Type{
+		"name": basetypes.StringType{},
+	}
+	coreAttrs := getPersistentGrantContract.CoreAttributes
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:951]: Is there any chance this could be nil?

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 9]
```diff
@@ -0,0 +1,1167 @@
...
+	// Test if values in sets match
+	matchNameBtwnScopes, matchVal := internaltypes.MatchStringInSets(scopeNames, eScopeNames)
+	if matchNameBtwnScopes {
+		resp.Diagnostics.AddError("Scope name conflict!", fmt.Sprintf("The scope name \"%s\" is already defined in another scope list", matchVal))
+	}
+}
+
+func addOptionalOauthAuthServerSettingsFields(ctx context.Context, addRequest *client.AuthorizationServerSettings, plan oauthAuthServerSettingsResourceModel) error {
+
+	if internaltypes.IsDefined(plan.DefaultScopeDescription) {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:652]: It looks like some of the attributes in this method such as this one aren't optional, so we don't need to re-add them to the addrequest here.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 10]
```diff
@@ -0,0 +1,1167 @@
...
+
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	apiReadOauthAuthServerSettings, httpResp, err := apiClient.OauthAuthServerSettingsApi.GetAuthorizationServerSettings(config.ProviderBasicAuthContext(ctx, providerConfig)).Execute()
+
+	if err != nil {
+		config.ReportHttpError(ctx, &resp.Diagnostics, "An error occurred while looking for a OauthAuthServerSettings", err, httpResp)
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:1082]: As I learned recently, we shouldn't report an error on read for a 404 https://developer.hashicorp.com/terraform/plugin/framework/resources/read#recommendations

The same applies to delete  https://developer.hashicorp.com/terraform/plugin/framework/resources/delete#recommendations

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 11]
```diff
@@ -0,0 +1,1167 @@
...
+	if err == nil {
+		tflog.Debug(ctx, "Read response: "+string(responseJson))
+	}
+	// Read the response
+	readOauthAuthServerSettingsResponse(ctx, updateOauthAuthServerSettingsResponse, &state)
+
+	// Update computed values
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:1148]: I realized these blocks were pointless recently so I removed them from the PD resources (as in the if statement -> return blocks at the very end of methods where they don't actually skip any logic)

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 12]
```diff
@@ -0,0 +1,1167 @@
...
+				Optional:    true,
+				PlanModifiers: []planmodifier.Int64{
+					int64planmodifier.UseStateForUnknown(),
+				},
+			},
+		},
+	}
+
+	// Set attributes in string list
+	if setOptionalToComputed {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/oauth_auth_server_settings_resource.go:583]: Looks like this is never going to run because the boolean is always false

**Review by henryrecker-pingidentity — CHANGES_REQUESTED** [review]

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 7 above)
Comment (henryrecker-pingidentity) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go:870]: Nevermind I see this is an edit-only resource based on the comment on the delete method.

File: Makefile [hunk 13]
```diff
@@ -52,7 +52,7 @@ testacc:
 	PINGFEDERATE_PROVIDER_USERNAME=administrator \
 	PINGFEDERATE_PROVIDER_PASSWORD=2FederateM0re \
 	PINGFEDERATE_PROVIDER_INSECURE_TRUST_ALL_TLS=true \
-	TF_ACC=1 go test -timeout 10m -v ./internal/... -p 4
+	TF_ACC=1 go test -timeout 20m -v ./internal/... -p 1
```
[the remark this answers is not in the document]
Comment (wesleymccollam) [reply Makefile:55]: Just wanted to match the PD provider

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 8 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go:951]: these are the values that are always returned from the client

File: Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (wesleymccollam) [reply Makefile:55]: Parallelism fixes tests that seem to conflict

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 6 above)
Comment (henryrecker-pingidentity) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: This one is fixed but there are similar ones below that start with defaults rather than reading straight from the plan

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go [hunk 14]
```diff
@@ -0,0 +1,1167 @@
...
+	schema := schema.Schema{
+		Description: "Manages Oauth Auth Server Settings",
+		Attributes: map[string]schema.Attribute{
+			"default_scope_description": schema.StringAttribute{
+				Description: "The default scope description.",
+				Required:    true,
+			},
+			"scopes": schema.SetNestedAttribute{
+				Description: "The list of common scopes.",
+				Computed:    true,
```
[the remark this answers is not in the document]
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go:111]: In the latest commit, all computed values were removed from the schema, tf apply, then added back as needed.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 6 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: It wasn't 🥇

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 9 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: Add unnecessary requests were removed in the latest commit.

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 10 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: 404 check added in the latest commit

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 11 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: now removed where applicable

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 12 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: doh! 👍 , removed

File: internal/resource/config/oauth/oauth_auth_server_settings_resource.go (hunk 6 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/oauth_auth_server_settings_resource.go]: the remaining properties set addRequest to use the type with defaults as properties within the property object are not nullable

**Review by henryrecker-pingidentity — APPROVED** [review]
